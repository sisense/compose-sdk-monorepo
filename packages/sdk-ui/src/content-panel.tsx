import styled from '@emotion/styled';
import { isPivotWidget, isTableWidget } from './dashboard-widget/utils';
import { Layout, WidgetModel } from './models';
import { ChartWidgetProps, PivotTableWidgetProps, TableWidgetProps } from './props';
import { ChartWidget } from './widgets/chart-widget';
import { PivotTableWidget } from './widgets/pivot-table-widget';
import { TableWidget } from './widgets/table-widget';

const DIVIDER_STYLE = '1px solid rgba(38, 46, 61, 0.1)';

const Row = styled.div<{ widths: number[] }>`
  display: grid;
  grid-template-columns: ${({ widths }) => widths.map((w) => `${w}%`).join(' ')};
  &:last-of-type {
    border-bottom: ${DIVIDER_STYLE};
  }
`;

const Column = styled.div`
  border-left: ${DIVIDER_STYLE};
  &:last-of-type {
    border-right: ${DIVIDER_STYLE};
  }
`;

const Subcell = styled.div<{ height: string | number }>`
  border-top: ${DIVIDER_STYLE};
  &:not(:first-of-type) {
    border-left: ${DIVIDER_STYLE};
  }
  height: ${({ height }) => `calc(${height} + 32px)`};
`;

function getWidgetProps(widgetModel: WidgetModel) {
  const { widgetType } = widgetModel;
  if (isTableWidget(widgetType)) {
    return widgetModel.getTableWidgetProps();
  } else if (isPivotWidget(widgetType)) {
    return widgetModel.getPivotTableWidgetProps();
  } else {
    return widgetModel.getChartWidgetProps();
  }
}

const renderWidgetModel = (w: WidgetModel | undefined) => {
  if (!w) {
    return null;
  }

  const chartProps = getWidgetProps(w);

  if (isTableWidget(w.widgetType)) {
    return <TableWidget {...(chartProps as TableWidgetProps)} />;
  } else if (isPivotWidget(w.widgetType)) {
    return <PivotTableWidget {...(chartProps as PivotTableWidgetProps)} />;
  } else {
    return <ChartWidget {...(chartProps as ChartWidgetProps)} />;
  }
};

/**
 * Props for the {@link ContentPanel} component.
 *
 * @internal
 */
export interface ContentPanelProps {
  /**
   * An object defining how the widgets should be laid out.
   */
  layout: Layout;

  /**
   * A list of widget models to render.
   */
  widgets: WidgetModel[];
}

/**
 * A React component used for rendering a layout of widgets.
 *
 * @example
 * Example of using the component with the useGetDashboardModel hook:
 * ```tsx
 * const { dashboard, isLoading, isError } = useGetDashboardModel({
 *   dashboardOid: 'dashboard-oid',
 *   includeWidgets: true,
 * });
 *
 * const { layout, widgets } = dashboard;
 * if (!layout || !widgets) return null;
 *
 * return (
 *   <ContentPanel
 *     layout={layout}
 *     widgets={widgets}
 *   />
 * );
 * ```
 * @param props - {@link ContentPanelProps}
 * @internal
 */
export const ContentPanel = ({ layout, widgets }: ContentPanelProps) => {
  return (
    <Row widths={layout.columns.map((c) => c.widthPercentage)}>
      {layout.columns.map((column, columnIndex) => (
        <Column key={columnIndex}>
          {column.cells.map((cell, cellIndex) => (
            <Row
              key={`${columnIndex},${cellIndex}`}
              widths={cell.subcells.map((sb) => sb.widthPercentage)}
            >
              {cell.subcells.map((subcell) => (
                <Subcell
                  key={`${subcell.widgetId},${subcell.widthPercentage}`}
                  height={subcell.height}
                >
                  {renderWidgetModel(widgets.find((w) => w.oid === subcell.widgetId))}
                </Subcell>
              ))}
            </Row>
          ))}
        </Column>
      ))}
    </Row>
  );
};
