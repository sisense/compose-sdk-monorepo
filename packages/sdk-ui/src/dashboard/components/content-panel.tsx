import { getDividerStyle } from '@/dashboard/utils';
import { WidgetsPanelLayout } from '@/models';
import { WidgetProps } from '@/props';
import styled from '@emotion/styled';
import { useThemeContext } from '@/theme-provider';
import { Widget } from '@/widgets/widget';
import isUndefined from 'lodash-es/isUndefined';
import isNumber from 'lodash-es/isNumber';

const SMALL_WIDTH = '600px';
const MEDIUM_WIDTH = '900px';

const GridContainer = styled.div<{ widths: number[] }>`
  display: grid;
  grid-template-columns: ${({ widths }) => widths.map((w) => `${w}%`).join(' ')};
  @container content-panel-container (max-width: ${SMALL_WIDTH}) {
    grid-template-columns: repeat(1, 1fr);
  }
  @container content-panel-container ((min-width: ${SMALL_WIDTH}) and (max-width: ${MEDIUM_WIDTH})) {
    grid-template-columns: ${({ widths }) =>
      widths.length > 1 ? 'repeat(2, 1fr)' : 'repeat(1, 1fr)'};
  }
`;

const Column = styled.div<{
  responsive?: boolean;
  colWidths: number[];
  dividerWidth: number;
  dividerColor: string;
}>`
  container-name: content-panel-column;
  container-type: ${({ responsive }) => (responsive ? 'inline-size' : 'unset')};
  &:not(:first-of-type) {
    border-left: ${({ dividerWidth, dividerColor }) => getDividerStyle(dividerColor, dividerWidth)};
  }
  @container content-panel-container ((min-width: ${SMALL_WIDTH}) and (max-width: ${MEDIUM_WIDTH})) {
    &:last-of-type {
      grid-column-end: ${({ colWidths }) => (colWidths.length % 2 === 0 ? 'span 1' : 'span 2')};
    }
  }
  @container content-panel-container (max-width: ${SMALL_WIDTH}) {
    &:not(:first-of-type) {
      border-left: none;
    }
  }
`;

const Row = styled.div<{ widths: number[] }>`
  display: grid;
  grid-template-columns: ${({ widths }) => widths.map((w) => `${w}%`).join(' ')};
  @container content-panel-column (max-width: ${SMALL_WIDTH}) {
    grid-template-columns: repeat(1, 1fr);
  }
  @container content-panel-column ((min-width: ${SMALL_WIDTH}) and (max-width: ${MEDIUM_WIDTH})) {
    grid-template-columns: ${({ widths }) =>
      widths.length === 1 ? 'repeat(1, 1fr)' : 'repeat(2, 1fr)'};
  }
`;

const Subcell = styled.div<{
  responsive?: boolean;
  rowWidths: number[];
  height?: string | number;
  dividerWidth: number;
  dividerColor: string;
}>`
  border-bottom: ${({ dividerWidth, dividerColor }) => getDividerStyle(dividerColor, dividerWidth)};
  &:not(:first-of-type) {
    border-left: ${({ dividerWidth, dividerColor }) => getDividerStyle(dividerColor, dividerWidth)};
  }

  height: ${({ height }) =>
    isUndefined(height)
      ? 'auto'
      : isNumber(height)
      ? `calc(${height}px + 32px)`
      : `calc(${height} + 32px)`};
  @container ((min-width: ${SMALL_WIDTH}) and (max-width: ${MEDIUM_WIDTH})) {
    &:last-of-type {
      grid-column-end: ${({ rowWidths }) => (rowWidths.length % 2 === 0 ? 'span 1' : 'span 2')};
    }
  }
  @container (max-width: ${SMALL_WIDTH}) {
    &:not(:first-of-type) {
      border-left: none;
    }
  }
`;

/**
 * Props for the {@link ContentPanel} component.
 *
 * @internal
 */
export interface ContentPanelProps {
  /**
   * An object defining how the widgets should be laid out.
   */
  layout: WidgetsPanelLayout;

  /**
   * If true adjust layout based on available width of content panel.
   *
   * If not specified, the default value is `false`.
   *
   * @internal
   */
  responsive?: boolean;

  /**
   * A list of widget props to render.
   */
  widgets: WidgetProps[];
}

/**
 * A React component used for rendering a layout of widgets.
 *
 * @param props - {@link ContentPanelProps}
 * @internal
 */
export const ContentPanel = ({ layout, responsive, widgets }: ContentPanelProps) => {
  const { themeSettings } = useThemeContext();

  const colWidths = layout.columns.map((c) => c.widthPercentage);

  return (
    <GridContainer widths={colWidths}>
      {layout.columns.map((column, columnIndex) => (
        <Column
          key={columnIndex}
          colWidths={colWidths}
          responsive={responsive}
          dividerWidth={themeSettings.dashboard.dividerLineWidth}
          dividerColor={themeSettings.dashboard.dividerLineColor}
        >
          {column.rows?.map((row, cellIndex) => {
            const rowWidths = row.cells.map((sb) => sb.widthPercentage);
            return (
              <Row key={`${columnIndex},${cellIndex}`} widths={rowWidths}>
                {row.cells.map((subcell) => {
                  const widgetProps = widgets.find((w) => w.id === subcell.widgetId);
                  if (
                    widgetProps?.widgetType === 'pivot' &&
                    subcell.height &&
                    widgetProps?.styleOptions?.isAutoHeight
                  ) {
                    widgetProps.styleOptions.isAutoHeight = false;
                  }
                  return (
                    <Subcell
                      key={`${subcell.widgetId},${subcell.widthPercentage}`}
                      responsive={responsive}
                      rowWidths={rowWidths}
                      height={subcell.height}
                      dividerWidth={themeSettings.dashboard.dividerLineWidth}
                      dividerColor={themeSettings.dashboard.dividerLineColor}
                    >
                      {widgetProps && <Widget {...widgetProps} />}
                    </Subcell>
                  );
                })}
              </Row>
            );
          })}
        </Column>
      ))}
    </GridContainer>
  );
};
