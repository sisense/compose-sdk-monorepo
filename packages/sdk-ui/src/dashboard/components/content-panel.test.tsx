import { WidgetProps } from '@/props';
import { render } from '@testing-library/react';
import { ContentPanel } from './content-panel';

vi.mock('@/widgets/widget', async () => {
  return { Widget: (props: WidgetProps) => <div data-testid={'widget'}>{props.id}</div> };
});

describe('ContentPanel', () => {
  it('should render widgets with provided layout', async () => {
    const widgets = [
      { id: 'widget-1' },
      { id: 'widget-2' },
      { id: 'widget-3' },
      { id: 'widget-4' },
    ] as WidgetProps[];
    const layout = {
      columns: [
        {
          widthPercentage: 70,
          rows: [
            {
              cells: [
                { widthPercentage: 60, widgetId: 'widget-1' },
                { widthPercentage: 40, widgetId: 'widget-2' },
              ],
            },
          ],
        },
        {
          widthPercentage: 30,
          rows: [
            {
              cells: [
                { widthPercentage: 25, widgetId: 'widget-3' },
                { widthPercentage: 75, widgetId: 'widget-4' },
              ],
            },
          ],
        },
      ],
    };
    const { container, getAllByTestId } = render(
      <ContentPanel widgets={widgets} layout={layout} />,
    );

    const [wrapperWidths, ...columnsWidths] = Array.from(
      container.querySelectorAll('[widths]'),
    ).map((el) => el.getAttribute('widths'));
    expect(wrapperWidths).toBe('70,30');
    expect(columnsWidths).toEqual(['60,40', '25,75']);

    const widgetElements = getAllByTestId('widget');
    expect(widgetElements).toHaveLength(4);
  });
});
