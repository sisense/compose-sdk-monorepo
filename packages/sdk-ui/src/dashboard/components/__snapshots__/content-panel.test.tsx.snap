// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContentPanel > should render widgets without provided layout 1`] = `
<div>
  <div
    class="css-k7yy55"
    widths="100"
  >
    <div
      class="css-1cpn397"
    >
      <div
        class="css-13i8pdy"
        widths="100"
      >
        <div
          class="css-18wpp0e"
        >
          <div
            data-testid="widget"
          >
            widget-1
          </div>
        </div>
      </div>
      <div
        class="css-13i8pdy"
        widths="100"
      >
        <div
          class="css-18wpp0e"
        >
          <div
            data-testid="widget"
          >
            widget-2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
